### ISSUE TITLE
可视化结果

### ISSUE THREAD

**Issue by Cherich-GL** [Shank2358/NPMMR-Det#9]
不好意思打扰您了，您能分享一下 MTR MSR NP 热图 以及在DIOR数据集上测试结果的实现方法吗  下面是我的邮箱
303256090@qq.com    如果可以的话万分感谢  谢谢~
![image](https://user-images.githubusercontent.com/95956260/162677459-5d969172-b339-47a2-b135-fd436b2fa6c4.png)
![image](https://user-images.githubusercontent.com/95956260/162677480-502983bc-78c8-4773-94f1-21d4e76d8250.png)

**Comment by Shank2358** [comment]
> 不好意思打扰您了，您能分享一下 MTR MSR NP 热图 以及在DIOR数据集上测试结果的实现方法吗 下面是我的邮箱 [303256090@qq.com](mailto:303256090@qq.com) 如果可以的话万分感谢 谢谢~ ![image](https://user-images.githubusercontent.com/95956260/162677459-5d969172-b339-47a2-b135-fd436b2fa6c4.png) ![image](https://user-images.githubusercontent.com/95956260/162677480-502983bc-78c8-4773-94f1-21d4e76d8250.png)

不客气呀，有问题随时可以交流。
1）第一个可视化结果的代码我加在这里了
![image](https://user-images.githubusercontent.com/33946139/162865870-1d231d9e-d708-4f62-bf55-f7d42f1eecba.png)
您可以单独封装一下它，我图省事直接写在循环里面了（类别和颜色数量要改成DIOR的即可）
2）第二个是画的特征图然后叠加到原图上，我这里写一个demo，您仿照着加到对应层即可
        ##导入必要的包
        import matplotlib.pyplot as plt
        import matplotlib.ticker as ticker
        import cv2
        #假设输入变量是x  (torch.Tensor类型)，维度是[Batch_Size, Channel, H, W]
        x_show = x.detach().cpu().permute(0, 2, 3, 1)  #先截断(.detach()) x的梯度，然后从GPU转到CPU，然后交换维度为[Batch_Size, H, W , Channel]
        x_show = torch.sigmoid(x1) #对特征归一化，我这里直接用sigmoid函数来，也可以用max-min
        heatmap = x_show[0, :, :, 0:1].numpy() #取batch的第一个图，可视化它的第一个通道，这个可以根据需求修改调整，得到heatmap，转成numpy数组
        heatmap = cv2.applyColorMap(heatmap, cv2.COLORMAP_JET) #调用CV2函数把单通道的heatmap转成彩图
        heatmap_on_img = cv2.addWeighted(img, 0.5, heatmap, 0.5, 0) #和原图叠加，权重是0.5和0.5，原图是img，记得大小要resize到和heatmap一致（用这个cv2.resize）
        plt.figure(’heatmap_on_img‘)  # 图像窗口名称
        plt.imshow(heatmap_on_img, cmap='jet') #画图，配色方案jet
        cb = plt.colorbar(fraction=0.05, pad=0.05) #画右边的颜色条
        cb.update_ticks()
        plt.show()
        ##后续还可以写保存等操作，我这儿就不展开了

可视化这个代码在另一个仓库里面有更多的，有需要也可以参考
![image](https://user-images.githubusercontent.com/33946139/162868173-74c7b756-6a2e-4f64-be4e-710d3b56a4cc.png)

谢谢，任何问题都可以给我留言，我会尽力帮忙~

**Comment by Cherich-GL** [comment]
很抱歉又要请教您，特征图叠加到原图上的可视化我还是没有成功。我在darknet53_npattention.py插入的您上面的demo
![image](https://user-images.githubusercontent.com/95956260/164718115-5e6ca08b-9146-4d60-990a-fd265e43623f.png)
出现 了以下的错误
![image](https://user-images.githubusercontent.com/95956260/164718700-2aaae342-80ce-4d84-b159-73ab8dff010c.png)
希望您能帮助我

**Comment by Shank2358** [comment]
> 很抱歉又要请教您，特征图叠加到原图上的可视化我还是没有成功。我在darknet53_npattention.py插入的您上面的demo ![image](https://user-images.githubusercontent.com/95956260/164718115-5e6ca08b-9146-4d60-990a-fd265e43623f.png) 出现 了以下的错误 ![image](https://user-images.githubusercontent.com/95956260/164718700-2aaae342-80ce-4d84-b159-73ab8dff010c.png) 希望您能帮助我

没事儿呀，有问题随时可以找我。
从目前来看好像是输入到cv2的数据类型不对，需要转换成uint8。
代码改成下面这样了，您再试试看：
        import matplotlib.pyplot as plt
        import cv2
        import numpy as np
        x_show = x0.detach().cpu().permute(0, 2, 3, 1)
        x_show = torch.sigmoid(torch.mean(x_show, dim=-1, keepdim=True))####我这里去了均值，也可以改成可视化某个通道
        heatmap = x_show[0, :, :, :].numpy()
        img1 = img.detach().cpu().permute(0, 2, 3, 1)
        img1 = (img1[0, :, :, :].numpy()*255).astype(np.uint8)
        heatmap = cv2.applyColorMap((heatmap*255).astype(np.uint8), cv2.COLORMAP_JET)
        plt.figure('image')
        plt.imshow(img1) #可视化原图
        ratio = 1 / 32  #图像下采样到和特征图一样大的下采样倍率
        img1 = cv2.resize(img1, dsize=None, fx=ratio, fy=ratio, interpolation=cv2.INTER_NEAREST)
        heatmap_on_img = cv2.addWeighted(img1, 0.7, heatmap, 0.3, 0)#########比例可调
        plt.figure('heatmap_on_img')  # 图像窗口名称
        plt.imshow(heatmap_on_img, cmap='jet')  # 画图，配色方案jet
        plt.colorbar()
        plt.show()
![image](https://user-images.githubusercontent.com/33946139/164728193-061a7c84-705f-4e1e-aec4-e8ddb19ab4c2.png)
跑出来长这样，这是还没训练时候的特征，训练完之后显著性区域会变红，和背景颜色区分开。

**Comment by Cherich-GL** [comment]
您这里的255是numpy/255吗
![image](https://user-images.githubusercontent.com/95956260/164879447-bbc15a70-6573-4acc-ada6-66d0aba166ea.png)
我是在npmmrdet_model.py里添加的这个模块，我把backbone输出的x_s作为的特征图  batch_size改成了1，numpy255改成了numpy/255，运行的train.py  出现了下面的结果
热图
![image](https://user-images.githubusercontent.com/95956260/164879587-aa0e44ff-c256-497f-bd77-ca6257afee69.png)
原图
![image](https://user-images.githubusercontent.com/95956260/164879602-0fe50bc5-6a61-4e7b-844d-6863833e925f.png)

**Comment by Cherich-GL** [comment]
![image](https://user-images.githubusercontent.com/95956260/164879624-2410d996-be8e-4a44-887f-85c164555bcc.png)
这是添加之后的forward

**Comment by Shank2358** [comment]
> 您这里的255是numpy/255吗 ![image](https://user-images.githubusercontent.com/95956260/164879447-bbc15a70-6573-4acc-ada6-66d0aba166ea.png) 我是在npmmrdet_model.py里添加的这个模块，我把backbone输出的x_s作为的特征图 batch_size改成了1，numpy255改成了numpy/255，运行的train.py 出现了下面的结果 热图 ![image](https://user-images.githubusercontent.com/95956260/164879587-aa0e44ff-c256-497f-bd77-ca6257afee69.png) 原图 ![image](https://user-images.githubusercontent.com/95956260/164879602-0fe50bc5-6a61-4e7b-844d-6863833e925f.png)

是*255，原来的范围是０－１，可视化要变成０－２５５，抱歉复制过来好像漏了

![image](https://user-images.githubusercontent.com/33946139/164887274-3ad4c34b-c983-49e5-8b37-397e24c35120.png)

colorbar如果有问题去掉看看

**Comment by Cherich-GL** [comment]
好的成功了，很感谢您的耐心解答。真心感谢大佬

**Comment by Shank2358** [comment]
> 好的成功了，很感谢您的耐心解答。真心感谢大佬

不客气，随时可以交流～～
还可以试试CAM等一些可视化的方法，画出来会更好看
